skfp/model_selection/splitters/pubchem_split.py

```diff
@@ -0,0 +1,456 @@
+
+    data_df = data_df[~data_df["year"].isna()]
+    data_df = data_df.sort_values("year")
+
+    train_size, valid_size, test_size = validate_train_valid_test_split_sizes(
+        train_size, valid_size, test_size, len(data)
+    )
+
+    train_idxs: list[int] = list(data_df.iloc[:train_size].idx)
+    valid_idxs: list[int] = list(data_df.iloc[train_size : train_size + valid_size].idx)
+    test_idxs: list[int] = list(data_df.iloc[train_size + valid_size :].idx)
+
+    if not_found_behavior == "train":
+        train_idxs.extend(not_found_mols)
+
+    if not_found_behavior == "test":
+        test_idxs.extend(not_found_mols)
+
+    if return_indices:
+        train_subset = train_idxs
+        valid_subset = valid_idxs
+        test_subset = test_idxs
+    else:
+        train_subset = get_data_from_indices(data, train_idxs)
+        valid_subset = get_data_from_indices(data, valid_idxs)
+        test_subset = get_data_from_indices(data, test_idxs)
+
+    ensure_nonempty_subset(train_subset, "train")
+    ensure_nonempty_subset(valid_subset, "validation")
+    ensure_nonempty_subset(test_subset, "test")
+
+    if additional_data:
+        additional_data_split: list[Sequence[Any]] = split_additional_data(
+            list(additional_data), train_idxs, valid_idxs, test_idxs
+        )
+        return train_subset, valid_subset, test_subset, *additional_data_split
+    else:
+        return train_subset, valid_subset, test_subset
+
+
+def _get_pubchem_years(
+    data: Sequence[Union[str, Mol]], n_jobs: int = 5
+) -> list[list[int]]:
+    """
+    Get first literature publication year from PubChem for a list of molecules, either
+    as SMILES strings or RDKit Mol objects.
+
+    Parameters
+    ----------
+    data : sequence
+        A sequence representing either SMILES strings or RDKit `Mol` objects.
+    """
+    data = list(
+        map(
+            lambda mol: (
+                MolToSmiles(mol, canonical=True, isomericSmiles=True)
+                if mol is Mol
+                else mol
+            ),
+            data,
+        )
+    )
+
+    print("Converting SMILES to CIDs")
+    cids = run_in_parallel(get_cid_for_smiles, data, n_jobs)
+
+    print("Searching for CIDs literature")
+    years = run_in_parallel(get_earliest_publication_date, cids, n_jobs)
+
+    return years
+
+
+def get_cid_for_smiles(smiles: str) -> Optional[str]:
+    """
+
+    Parameters:
+    ------------
+    smiles: str
+        SMILES string
+
+    Return
+    --------
+        PubChem compound identifier if  else None
+
+    """
+    PUG_url = f"https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/smiles/{quote(smiles)}/cids/JSON"
+
+    PUG_response = requests.get(PUG_url, timeout=10)
+    PUG_response_status = PUG_response.status_code
+
+    if PUG_response_status != 200:
+        warnings.warn(
+            f"PUG error \n, {PUG_url} \n status code {PUG_response_status}, \n check your input"
+        )
```

Comment:
PUG can easily return errors on server overloaded error. This has to be resilient and implement retries. They may also be configurable in the splitter, e.g. 10 retries by default, with 1 second wait in between.